crates/m3/src/builder/indexed_lookup.rs

```diff
@@ -0,0 +1,431 @@
+			.map(|val| Boundary {
+				values: vec![B32::new(val).into()],
+				direction: FlushDirection::Pull,
+				channel_id: incr_lookup_chan,
+				multiplicity: 1,
+			})
+			.collect::<Vec<_>>();
+
+		// Tally the lookup counts from the looker tables
+		let counts =
+			tally(&cs, &mut witness, &boundaries, incr_lookup_chan, &IncrIndexedLookup).unwrap();
+
+		// Fill the lookup table with the sorted counts
+		let sorted_counts = counts
+			.into_iter()
+			.enumerate()
+			.sorted_by_key(|(_, count)| Reverse(*count))
+			.collect::<Vec<_>>();
+
+		witness
+			.fill_table_sequential(&incr_lookup, &sorted_counts)
+			.unwrap();
+
+		validate_system_witness::<OptimalUnderlier>(&cs, witness, boundaries);
+	}
+
+	fn merge_incr_cols(
+		table: &mut TableBuilder,
+		input: Col<B8>,
+		carry_in: Col<B1>,
+		output: Col<B8>,
+		carry_out: Col<B1>,
+	) -> Col<B32> {
+		let beta_1 = ext_basis::<B32, B8>(1);
+		let beta_2_0 = ext_basis::<B32, B8>(2);
+		let beta_2_1 = beta_2_0 * ext_basis::<B8, B1>(1);
+		table.add_computed(
+			"merged",
+			upcast_col(input)
+				+ upcast_col(output) * beta_1
+				+ upcast_col(carry_in) * beta_2_0
+				+ upcast_col(carry_out) * beta_2_1,
+		)
+	}
+
+	fn merge_incr_vals(input: u8, carry_in: bool, output: u8, carry_out: bool) -> u32 {
+		((carry_out as u32) << 17)
+			| ((carry_in as u32) << 16)
+			| ((output as u32) << 8)
+			| input as u32
+	}
+
+	struct Incr {
+		pub input: Col<B8>,
+		pub carry_in: Col<B1>,
+		pub output: Col<B8>,
+		pub carry_out: Col<B1>,
+		pub merged: Col<B32>,
+	}
+
+	impl Incr {
+		fn new(
+			table: &mut TableBuilder,
+			lookup_chan: ChannelId,
+			input: Col<B8>,
+			carry_in: Col<B1>,
+		) -> Self {
+			let output = table.add_committed::<B8, 1>("output");
+			let carry_out = table.add_committed::<B1, 1>("carry_out");
+			let merged = merge_incr_cols(table, input, carry_in, output, carry_out);
+
+			table.pull(lookup_chan, [merged]);
+
+			Self {
+				input,
+				carry_in,
+				output,
+				carry_out,
+				merged,
+			}
+		}
+
+		fn populate<P>(&self, witness: &mut TableWitnessSegment<P>) -> anyhow::Result<()>
+		where
+			P: PackedFieldIndexable<Scalar = B128>
+				+ PackedExtension<B1>
+				+ PackedExtension<B8>
+				+ PackedExtension<B32>,
+		{
+			let input = witness.get_as::<u8, _, 1>(self.input)?;
+			let carry_in = witness.get(self.carry_in)?;
+			let mut output = witness.get_mut_as::<u8, _, 1>(self.output)?;
+			let mut carry_out = witness.get_mut(self.carry_out)?;
+			let mut merged = witness.get_mut_as::<u32, _, 1>(self.merged)?;
+
+			for i in 0..witness.size() {
+				let input_i = input[i];
+				let carry_in_bit = bool::from(get_packed_slice(&carry_in, i).val());
+
+				let (output_i, carry_out_bit) = input_i.overflowing_add(carry_in_bit.into());
+				output[i] = output_i;
+				set_packed_slice(&mut carry_out, i, B1::from(carry_out_bit));
+				merged[i] = ((carry_out_bit as u32) << 17)
+					| ((carry_in_bit as u32) << 16)
+					| ((output_i as u32) << 8)
+					| input_i as u32;
```

Comment:
Wait innit this merge_incr_vals?

Reply:
Yep, it is. This is fixed in the next change in the stack. OK to leave it as is in this one?